runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -0,0 +1,436 @@
+    public V put(@ConflictParameter K key, V value) {
+        V previous = mainMap.put(key, value);
+        // If we have index functions, update the secondary indexes.
+        if (indexFunctions.size() > 0) {
+            unmapSecondaryIndexes(key, previous);
+            mapSecondaryIndexes(key, value);
+        }
+        return previous;
+    }
+
+    @DontInstrument
+    protected V undoPutRecord(CorfuTable<K, V, F, I> table, K key, V value) {
+        return table.mainMap.get(key);
+    }
+
+    @DontInstrument
+    protected void undoPut(CorfuTable<K, V, F, I> table, V undoRecord, K key, V value) {
+        // Same as undoRemove (restore previous value)
+        undoRemove(table, undoRecord, key);
+    }
+
+    @DontInstrument
+    Object[] putAllConflictFunction(Map<? extends K, ? extends V> m) {
+        return m.keySet().stream()
+                .map(Object::hashCode)
+                .toArray(Object[]::new);
+    }
+
+    enum UndoNullable {
+        NULL;
+    }
+
+    /** Generate an undo record for putAll, given the previous state of the map
+     * and the parameters to the putAll call.
+     *
+     * @param previousState     The previous state of the map
+     * @param m                 The map from the putAll call
+     * @return                  An undo record, which for a putAll is all the
+     *                          previous entries in the map.
+     */
+    @DontInstrument
+    @SuppressWarnings("unchecked")
+    Map<K,V> undoPutAllRecord(CorfuTable<K, V, F, I> previousState,
+                              Map<? extends K, ? extends V> m) {
+        ImmutableMap.Builder<K,V> builder = ImmutableMap.builder();
+        m.keySet().forEach(k -> builder.put(k,
+                (previousState.get(k) == null
+                        ? (V) CorfuTable.UndoNullable.NULL
+                        : previousState.get(k))));
+        return builder.build();
+    }
+
+    /** Undo a remove, given the current state of the map, an undo record
+     * and the arguments to the remove command to undo.
+     *
+     * @param table           The state of the map after the put to undo
+     * @param undoRecord    The undo record generated by undoRemoveRecord
+     */
+    @DontInstrument
+     void undoPutAll(CorfuTable<K, V, F, I> table, Map<K,V> undoRecord,
+                            Map<? extends K, ? extends V> m) {
+        undoRecord.entrySet().forEach(e -> {
+                    if (e.getValue() == CorfuTable.UndoNullable.NULL) {
+                        undoRemove(table, null, e.getKey());
+                    } else {
+                        undoRemove(table, e.getValue(), e.getKey());
+                    }
+                }
+        );
+    }
+
+    @Mutator(name = "put", noUpcall = true)
+    void insert(@ConflictParameter K key, V value) {
+        V previous = mainMap.put(key, value);
+        // If we have index functions, update the secondary indexes.
+        if (indexFunctions.size() > 0) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Use isEmpty() to check whether the collection is empty or not. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1155)